### PR TITLE
kernel/load_self: implement late binding for export variables

### PR DESCRIPTION
**Pull request by bookmist**
This PR should fix crashes in games with "[load_var_imports]: 	NID NOT FOUND" warning. It's almost all unity games and some other. I check on "miku miku hockey" game.

The code looks not so good organized, so if you have suggestion of how to make it more simple and understandable, please, comment.

### CODE REVIEW

File: vita3k/modules/module_parent.cpp [hunk 1]
```diff
@@ -168,8 +168,7 @@ void call_import(EmuEnvState &emuenv, CPUState &cpu, uint32_t nid, SceUID thread
         const std::unordered_set<uint32_t> lle_nid_blacklist = {};
         log_import_call('L', nid, thread_id, lle_nid_blacklist, pc);
         write_pc(cpu, export_pc);
-        // TODO: invalidate cache for all threads. Now invalidate_jit_cache is not thread safe.
-        invalidate_jit_cache(cpu, pc, 4 * 3);
+        emuenv.kernel.invalidate_jit_cache(pc, 4 * 3);
```
Comment (Macdu) [inline vita3k/modules/module_parent.cpp:171]: Isn't this function not thread safe?

File: vita3k/kernel/src/load_self.cpp [hunk 2]
```diff
@@ -70,13 +70,19 @@ static bool load_var_imports(const uint32_t *nids, const Ptr<uint32_t> *entries,
 
         VarImportsHeader *const var_reloc_header = reinterpret_cast<VarImportsHeader *>(entry.get(mem));
         const auto var_reloc_entries = static_cast<void *>(var_reloc_header + 1);
+        const uint32_t reloc_size = (var_reloc_header->reloc_data_size > sizeof(VarImportsHeader)) ? (var_reloc_header->reloc_data_size - sizeof(VarImportsHeader)) : 0;
 
         Address export_address;
         const ExportNids::iterator export_address_it = kernel.export_nids.find(nid);
```
Comment (Macdu) [inline vita3k/kernel/src/load_self.cpp:76]: This was already a fault in the code but can you fix it? the call to `export_nids.find` should be with a shared_lock (it is not thread safe).

File: vita3k/kernel/src/load_self.cpp [hunk 3]
```diff
@@ -70,13 +70,19 @@ static bool load_var_imports(const uint32_t *nids, const Ptr<uint32_t> *entries,
...
+        const char *const name = import_name(nid);
         if (export_address_it != kernel.export_nids.end()) {
             export_address = export_address_it->second;
+            if (kernel.late_binding_infos.contains(nid)) {
```
Comment (Macdu) [inline vita3k/kernel/src/load_self.cpp:80]: The call to contains is not thread safe, putting the lock_exclusive above is the easiest way to fix this.

File: vita3k/kernel/src/load_self.cpp [hunk 4]
```diff
@@ -263,13 +270,68 @@ static bool load_var_exports(const uint32_t *nids, const Ptr<uint32_t> *entries,
             LOG_DEBUG("\tNID {} ({}) at {}", log_hex(nid), name, log_hex(entry.address()));
         }
 
-        {
+        if (!kernel.late_binding_infos.contains(nid)) {
             const std::unique_lock<std::shared_mutex> lock(kernel.export_nids_mutex);
             kernel.export_nids.emplace(nid, entry.address());
+        } else {
+            LOG_DEBUG("Found previously not found variable. nid:{}, new_entry_point:{}", log_hex(nid), log_hex(entry.address()));
+            Address old_entry_address = kernel.export_nids[nid];
```
Comment (Macdu) [inline vita3k/kernel/src/load_self.cpp:278]: This is not thread safe, put it under the lock_exclusive

File: vita3k/modules/module_parent.cpp (hunk 1 above)
Comment (bookmist) [reply vita3k/modules/module_parent.cpp]: Sorry, I forgot to revert it. I was trying to make this function thread safe but was not successful. Now fixed.

File: vita3k/kernel/src/load_self.cpp [hunk 5]
```diff
@@ -600,6 +657,7 @@ SceUID load_self(Ptr<const void> &entry_point, KernelState &kernel, MemState &me
     const SceUID uid = kernel.get_next_uid();
     sceKernelModuleInfo->modid = uid;
     kernel.loaded_modules.emplace(uid, sceKernelModuleInfo);
+    kernel.module_uid_by_nid.emplace(module_info->module_nid, uid);
```
Comment (Macdu) [inline vita3k/kernel/src/load_self.cpp:660]: This call is not thread safe (module_uid_by_nid is synchronised using export_nids_mutex and not kernel.mutex)

Comment (bookmist) [reply vita3k/kernel/src/load_self.cpp]: Fixed. thank you.

**Review by Macdu — APPROVED** [review]
LGTM

**Pull request merged**